### PR TITLE
Add `has_lib` and `bin_names` fields to the versions API

### PR DESCRIPTION
**Pull request by Turbo87**
https://docs.rs/cargo-manifest/latest/cargo_manifest/struct.Manifest.html#method.complete_from_abstract_filesystem is used to find implicit libraries or binaries in the crate file.

Tasks:

- [x] find libraries, binaries, etc. in the crate file
- [x] save that data in the database
- [x] expose that data in the API
- [x] adjust the install instructions on the web frontend depending on the data (see https://github.com/rust-lang/crates.io/pull/8883)
- [ ] figure out a way to backfill the data for existing versions

Related:

- https://github.com/rust-lang/crates.io/issues/5882
- https://github.com/rust-lang/crates.io/pull/8883

### CODE REVIEW

**Comment by Turbo87** [comment]
This PR is ready for review now. The frontend part of this work will be opened in an independent PR, same for the backfilling code.

File: migrations/2024-06-14-081653_add-lib-and-bins-columns/up.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+alter table versions
+    add has_lib boolean,
+    add bin_names text[];
+
+comment on column versions.has_lib is 'TRUE if the version has a library (e.g. `src/lib.rs`), FALSE if no library was detected, or NULL if the version has not been analyzed yet.';
+comment on column versions.bin_names is 'list of the names of all detected binaries in the version. the list may be empty which indicates that no binaries were detected in the version. the column may be NULL is the version has not been analyzed yet.';
```
Comment (eth3lbert) [inline migrations/2024-06-14-081653_add-lib-and-bins-columns/up.sql:6]: IIUC, null is used for existing versions that need backfilling? And for newly published versions, a non-null value would be expected to be inserted? Do we also plan to alter the table to non-nullable once all values are filled?

Comment (Turbo87) [reply migrations/2024-06-14-081653_add-lib-and-bins-columns/up.sql:6]: > IIUC, null is used for existing versions that need backfilling? And for newly published versions, a non-null value would be expected to be inserted?

correct

> Do we also plan to alter the table to non-nullable once all values are filled?

I'm not sure yet. at the moment the detection is done inside the publish endpoint, but if we were to move the detection to an async background worker job we might want to keep this "no-yet-analyzed" state.

**Review by LawnGnome — APPROVED** [review]
With the obvious caveat that we're now dependent on `cargo-manifest`'s target discovery working the same way as `cargo`, this LGTM. (On the bright side, this is something that's easily re-backfilled later if there are any issues, and certainly not on any critical path for publishing or using crates.)

**Comment by Turbo87** [comment]
> With the obvious caveat that we're now dependent on `cargo-manifest`'s target discovery working the same way as `cargo`, this LGTM.

yeah, that's why I'm currently working on fixing some of the finer details of the target auto-discovery in `cargo-manifest` :D

File: migrations/2024-06-14-081653_add-lib-and-bins-columns/up.sql (hunk 1 above)
Comment (eth3lbert) [reply migrations/2024-06-14-081653_add-lib-and-bins-columns/up.sql:6]: Thanks for the explanation!

This actually reminds me of `default_versions`, which faces a similar situation but actually handles it quite well.

**Comment by Turbo87** [comment]
https://github.com/rust-lang/crates.io/pull/8951 has been merged, so this is now unblocked :)

**Pull request merged**